### ISSUE TITLE
[Milestone] v4

### ISSUE THREAD

**Issue by jcbhmr** [Andrew-Chen-Wang/github-wiki-action#30]
I've recently made [a lot of changes](https://github.com/Andrew-Chen-Wang/github-wiki-action/tree/jcbhmr) that I'd like to incorporate into a v4 release. There's no risk of breaking changes because **nobody uses the `@master` tag**! #23 It's awesome that good versioning practices are so prevalent! ❤️

Here's a summary of some of the changes that I'd like to incorporate into this repo:
- [x] Use `with:` instead of `env:` like all other actions #27 
- [x] No longer using rsync to put files into a git repo. Instead, we use $GIT_WORK_TREE to move the git worktree to the files, not the files to the git worktree! 😲 #31
- [x] Use a non-Docker based method for building the action. #32 
- [x] Add a test.yml GitHub workflow to make sure that the action works in PRs! #33 
- [x] #34
- [x] #35
- [x] Use a `strategy: clone|init` switch to either a) add a commit to the existing repo or b) create a single commit and force push that. #36
  - Some people might want the semantics of a single deployed commit instead of a continuously stacking uncontrolled git history. More options are better.
  - For large image blobs (popular in some auto-generated wikis!) the git history gets bloated by so much non-diff-able `.jpg` 10MB files that it becomes slow over time. Force pushing a single commit solves this.
- [x] Keeping `strategy: clone` as the default is perfect.
- [x] #37
- [x] Use `gh auth setup-git` to manage git credentials. #38
  - Dealing with tokens seems complicated. Delegating to an official tool that just needs a `GITHUB_TOKEN` var set is great!
  - `gh` is ALWAYS available on runners. Windows, macOS, and Ubuntu.
- [x] Add a note in the readme about #9 #8 #26
  - Preprocessing links to make them accurate in both GitHub source view and GitHub wiki is a good idea! It's a great fit for another action. And there's already an action that fits this: https://github.com/marketplace/actions/strip-markdown-extensions-from-links-action
- [x] Added an update-tags.yml action to auto-create `v4` tags from `v4.0.0` releases and keep them always updated. There's an official action for this! When you release a `vX.Y.Z` tag it changes `vX` to point to it. #39

**Comment by jcbhmr** [comment]
I know this is like 2 months late. 😬 But hey, better late than never!

**Comment by jcbhmr** [comment]
sidenote: consider renaming `master` to `main` to be "hip" and "cool" 😆 https://github.com/github/renaming

**Comment by jcbhmr** [comment]
todo: @jcbhmr -- add the improvements from the readme https://github.com/Andrew-Chen-Wang/github-wiki-action/blob/jcbhmr/README.md to this list to do and track issues and open prs for that stuff

**Comment by Andrew-Chen-Wang** [comment]
@jcbhmr Thanks so much for continuing this work. Since you've managed to make things backwards compatible, I've opened the former rule up such that now you can open PRs and merge them yourself or commit directly to the base branch.

I think we should refrain from moving to main for now since some users are using `master` as the target version. Perhaps a year down the line.

**Comment by jcbhmr** [comment]
> I think we should refrain from moving to main for now since some users are using master as the target version. Perhaps a year down the line.

No worries! ❤️ I wouldn't worry too much about `Andrew-Chen-Wang/github-wiki-action@master` though; there doesn't _appear_ to be any **public** usage of it ([github search](https://github.com/search?q=Andrew-Chen-Wang%2Fgithub-wiki-action%40master&type=code)) ([sourcegraph search](https://sourcegraph.com/search?q=context%3Aglobal+Andrew-Chen-Wang%2Fgithub-wiki-action%40master&patternType=standard&sm=1&groupBy=repo)). (I don't know about private) #23

**Comment by jcbhmr** [comment]
> @jcbhmr Thanks so much for continuing this work. Since you've managed to make things backwards compatible, I've opened the former rule up such that now you can open PRs and merge them yourself or commit directly to the base branch.

I want to caution that "backwards compatible" might not be the right word. I've tried to stick with the core idea of "push a folder to the .wiki.git repo", but I've done so by:
1. Changing from `env:` to `with:`
2. Adding _more_ options
3. Changing the names of some of the options (`WIKI_DIR` => `path:`, `REPO` => `repository:`, etc)
4. Changing the core architecture of this project (Docker+Bash => Bash)
5. Using `GIT_WORK_TREE` and `GIT_DIR` instead of rsync

These changes, while good and (IMO) improvements, are nonetheless **a new version** and require **changing a config**. IDK if that's "backwards compatible" by the definition? 🤷‍♂️

```yml
      - uses: Andrew-Chen-Wang/github-wiki-action@v3
        # ⚠️ We use the env: key to provide our inputs! See #27.
        env:
          # Make sure this has a slash at the end! We use wiki/ by default.
          WIKI_DIR: my-octocat-wiki/
          # You MUST manually pass in the GitHub token.
          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          # These are currently REQUIRED options
          GH_MAIL: actions@users.noreply.github.com
          GH_NAME: actions[bot]
```

to

```yml
      - uses: Andrew-Chen-Wang/github-wiki-action@v4
        with:
          path: my-octocat-wiki
          token: ${{ secrets.MY_TOKEN }}
        env:
          GIT_AUTHOR_NAME: Myself
          GIT_AUTHOR_EMAIL: me@example.com
```

**hence**, the `v3` => `v4` upgrade version bump. Which, again, is recommended practice! Bumping versions on improvements to the core idea of the project **is good** and that's what version tags are for, right? 😜

---

TLDR: I want explicit ✅ that going from (left) to (right) is 👍 for v4. Just to make sure we are on the same page.

<table><td>

![image](https://github.com/Andrew-Chen-Wang/github-wiki-action/assets/61068799/b5558a01-4eb7-4349-bff4-f5bd38c72aac)

<td>

![image](https://github.com/Andrew-Chen-Wang/github-wiki-action/assets/61068799/6b54dab5-bf31-4179-a314-1f8f6d8ade86)

</table>

I plan on merging all the PRs and then @-pinging you about making a v4.0.0 release. Sound good?
#40  #41  #42  #43  #44  #45  #47

**Comment by Andrew-Chen-Wang** [comment]
Yes this sounds good. Major version update sounds good. Can't remember if the PR for the README includes a header saying "Breaking Changes in v4" which can help people migrate. I noticed a migration section iirc so that's great.

Plan sounds good to me. Thanks again.

**Comment by jcbhmr** [comment]
current status: it works as of 070db4423e74902abf6480a811234f81c64239da

![image](https://github.com/Andrew-Chen-Wang/github-wiki-action/assets/61068799/af35deb1-e850-4c4d-98cd-87a9d1eeb696)
![image](https://github.com/Andrew-Chen-Wang/github-wiki-action/assets/61068799/24f6689f-aa7d-474c-9075-ae25a592045f)

**Comment by jcbhmr** [comment]
@Andrew-Chen-Wang All done!
You can see it work here: 🥳

![image](https://github.com/Andrew-Chen-Wang/github-wiki-action/assets/61068799/d5021aa2-1373-4611-98cf-e6d1bccaefa2)
![image](https://github.com/Andrew-Chen-Wang/github-wiki-action/assets/61068799/332ff633-67ed-47f7-b4f7-73a2998c2e5b)

Here's what I need from you...
1. **create a GitHub Release**
2. use the name & tag v4.0.0 -- don't worry! the update-tags action will do the v4 tag for you!
3. click the "Generate release notes" button to use PRs to generate release notes

![image](https://github.com/Andrew-Chen-Wang/github-wiki-action/assets/61068799/481879af-e0c7-4bdf-9cfc-5bbe42bf215a)

You can also include this text in the release textbox copy+paste:
```md
✈️ If you're using v3, check out our [v3 ➡️ v4 migration guide](https://github.com/Andrew-Chen-Wang/github-wiki-action/discussions/50)
``

then you _could_ also:
1. change the repo description to match the readme slogan #46 `📖 GitHub Action to sync a folder to the GitHub wiki`
2. change the repo tags to not include `actions-docker`
3. if you could, move the #50 discussion to the "Announcements" section of the discussions (I can't since I don't have perms to do that)

**Comment by jcbhmr** [comment]
@Andrew-Chen-Wang All done! It works! I'm drafting Release v4.0.0 right now! I also need to actually write a complete guide: #50

![image](https://github.com/Andrew-Chen-Wang/github-wiki-action/assets/61068799/511eceae-b0b4-47e4-a9fe-2bd816eb1d0c)

**Comment by jcbhmr** [comment]
Here's the draft release.

![image](https://github.com/Andrew-Chen-Wang/github-wiki-action/assets/61068799/445655fd-58ec-4c4d-8a14-45c885a03de8)

I will hit "Go" when I finish writing #50

**Comment by jcbhmr** [comment]
Actually scratch the idea of #50 with a migration guide announcement post. Instead, since it's so short, why not put that in the release post? 🤔 That's what release posts are all for after all, right? 😜

**Comment by jcbhmr** [comment]
@Andrew-Chen-Wang Created release! https://github.com/Andrew-Chen-Wang/github-wiki-action/releases/tag/v4.0.0

🎉🎉🎉🎉

**Comment by Andrew-Chen-Wang** [comment]
awesome thanks again!
